implement TCP & partial windows support

xeijin opened the pull request:
taking an initial stab at this. Haven't got a windows machine handy to test right now, but confirmed this does work on macos when setting `server-use-tcp` to `'t`.

closes #1

Commit:
implement TCP server

xeijin commented:
Example of default output with two servers running, one via TCP and one via Socket (note addition of `-server-file`)

```
e evaluates a lisp expression and prints the result.
usage: e {args} expression {expressions...}
  -q	Shorthand for --quoted
  -quoted
    	Keep quoted strings.
  -server-file string
    	Emacs server TCP file (default "/Users/xeijin/.config/emacs/server/server")
  -socket-name string
    	Emacs server unix socket (default "/var/folders/wy/391mbt1905b9djsb2wg9yrbc0000gn/T/emacs501/server")
```

xeijin committed:
Update README.md

xeijin committed:
Update README.md

xeijin committed:
Update README.md

Commit:
improve test coverage

Commit:
handle serverFile  for fifo

Commit:
Merge branch 'master' of https://github.com/xeijin/emacsclient-commands

Commit:
updates to tests

required due to increased usage of `checkPath`

Commit:
update function name for consistency

Commit:
improve environment variable resets

Commit:
implement windows support

xeijin committed:
First working windows version

e.go now builds successfully (and works), but some other cmds still fail

Commit:
restore SendTTY

Commit:
restore SendTTY

szermatt commented on client.go:
```diff
@@ -37,8 +50,85 @@ func defaultSocketName() string {
+
+	xdgPathA := path.Join(os.Getenv("XDG_CONFIG_HOME"), emacsDirName)
+	xdgPathB := path.Join(os.Getenv("HOME"), ".config", emacsDirName) // user following convention without XDG_CONFIG_HOME set
+	legacyPath := path.Join(os.Getenv("HOME"), emacsDirName)          // windows (if HOME is set) and emacs pre-v27
```
The directory name should always be ".emacs.d" both on Unix and on Windows

xeijin replied:
I think it could be either, but I believe it defaults to  `~/.config/emacs` if you are following the `XDG` specification (which is most linux distros now?)

szermatt replied:
Sorry, I meant to only highlight the last line. 

It should be:

xdgPathA := path.Join(os.Getenv("XDG_CONFIG_HOME"), emacsDirName)
xdgPathB := path.Join(os.Getenv("HOME"), ".config", emacsDirName) 
legacyPath := path.Join(os.Getenv("HOME"), ".emacs.d")

szermatt commented on client.go:
```diff
@@ -37,8 +50,85 @@ func defaultSocketName() string {
+
+	fp, err := os.Open(serverFile)
+	if err != nil {
+		panic(err)
```
It would be best to handle this case. Nothing guarantees that opening the server file won't fail.

szermatt commented on client.go:
```diff
@@ -37,8 +50,85 @@ func defaultSocketName() string {
 func Dial(options *Options) (net.Conn, error) {
+	if checkPath(options.SocketName) {
+		return dialUnix(options)
+	} else {
```
It would be a good idea to check here that the serverPath exists and if it doesn't return an error that says that no emacs server is available.

szermatt reviewed (changes requested):
Thank you for making the changes! I can't try it on Windows, but it did work on Linux with a server file env variable set.

However, the first time I tried running a command with this patch in and no servers on Linux, I got a puzzling panic instead of an error message. I left some suggestions to help address this issue. Not having any Emacs server running is a common scenario, this is why I'm not merging this right away. 

I also left a suggestion to fix the default server file on Unix.

Commit:
handle absent server file & refactor tests

It would be a good idea to check here that the serverPath exists and if it doesn't return an error that says that no emacs server is available.

_Originally posted by @szermatt in https://github.com/szermatt/emacsclient-commands/pull/3#discussion_r624732360_

Commit:
handle absent server file & refactor tests

It would be a good idea to check here that the serverPath exists and if it doesn't return an error that says that no emacs server is available.

_Originally posted by @szermatt in https://github.com/szermatt/emacsclient-commands/pull/3#discussion_r624732360_

Commit:
Merge branch 'master' of https://github.com/xeijin/emacsclient-commands

Commit:
refactor test case 'no server file present'

xeijin committed:
idiomatic path notaiton for windows

xeijin committed:
remove redundant logic

xeijin committed:
skip emerge on windows for now

need to find a windows friendly alternative to SIGUSR1

xeijin committed:
workaround test failure due to bug on windows

xeijin committed:
create a make file for windows

ebuf.go is skipped due to failing fifo test -- need to debug

emerge.go is skipped due lack of windows alternative to SIGUSR1

xeijin commented:
@szermatt thanks for the comments, have responded in kind

I found a couple of other issues on Windows: 
- the `fifo` implementation for windows does not appear to be working (test fails) - this will need further debugging - this breaks `ebuf.go`
- I haven't found an straightforward answer to the lack of `SIGUSR1` on windows, this means `emerge.go` does not compile

Not being a `make` expert,  I added crude alternative for windows: `make.bat` - which currently skips both of these issues - otherwise Windows/TCP support is mostly there, so perhaps worth merging in for now and I can raise an issue for these two?

Commit:
addressing paths issue

https://github.com/szermatt/emacsclient-commands/pull/3/files#r624731680

Commit:
addressing paths issue

https://github.com/szermatt/emacsclient-commands/pull/3/files#r624731680

Commit:
Merge branch 'master' of https://github.com/xeijin/emacsclient-commands

Commit:
add missing return

xeijin committed:
Merge branch 'szermatt:master' into master

xeijin committed:
remove unneeded import declaration

szermatt commented:
Sorry for the long silence! It looks like I missed quite some updates in May. You've been busy :) 

> I found a couple of other issues on Windows:
> 
> the fifo implementation for windows does not appear to be working (test fails) - this will need further debugging - this breaks ebuf.go

Yes, I very much doubt this approach can work on windows without a custom re-implementation. However, using a fifo is just an optimisation: it is possible to make ebuf communicate through the socket or TCP and send the data as strings through that channel, chunk by chunk.  Such an approach would work on windows and unix; it would just be a bit slower. 

Using a fifo could be an option on unix and turned off on windows.

> I haven't found an straightforward answer to the lack of SIGUSR1 on windows, this means emerge.go does not compile

Here, too, it should be possible to make it work without it , just a little bit worse: it would be possible to ask the user to press return when they're done, which would trigger command on the emacs-side that checks whether merging is done and continue. 

I'm sure there's a way of doing IPC on Windows to replace signals, but I have to admit I have zero experience of Windows development, so I can't help with that.

> Not being a make expert, I added crude alternative for windows: make.bat - which currently skips both of these issues - otherwise Windows/TCP support is mostly there, so perhaps worth merging in for now and I can raise an issue for these two?

I have nothing against using make.bat under Windows; as long as it works. It should be possible to make make work under both; it seems that Windows sets an env variable "OS" that could be checked. 

Yes, this looks definitely worth merging. Last time, I didn't merge because it affected normal start, but now it only improves the commands, even if support on Windows is not complete yet.

szermatt commented:
It's merged. Thank you very much for the feature, and sorry for the long silence.

Feel free to send out more pull request if you'd like to make any follow-ups changes !

szermatt commented:
One question about TestParseServerFile in client_test.go

```
	var serverFileContents string = `127.0.0.1:62989 17061
 ;\I^|/+?<egxc[7Qb;6vGCp2:~6nhzcP>:8W#u&*}:@GJj&;ib5KU+).2N}S9Y(e%`
```

Do you know what this string means or is it just a capture ?

xeijin commented:
> One question about TestParseServerFile in client_test.go
> 
> ```
> 	var serverFileContents string = `127.0.0.1:62989 17061
>  ;\I^|/+?<egxc[7Qb;6vGCp2:~6nhzcP>:8W#u&*}:@GJj&;ib5KU+).2N}S9Y(e%`
> ```
> 
> Do you know what this string means or is it just a capture ?

It's the TCP auth key that emacs generates with `server-get-auth-key`

xeijin commented:
> Sorry for the long silence! It looks like I missed quite some updates in May. You've been busy :)
> 
> > I found a couple of other issues on Windows:
> > the fifo implementation for windows does not appear to be working (test fails) - this will need further debugging - this breaks ebuf.go
> 
> Yes, I very much doubt this approach can work on windows without a custom re-implementation. However, using a fifo is just an optimisation: it is possible to make ebuf communicate through the socket or TCP and send the data as strings through that channel, chunk by chunk. Such an approach would work on windows and unix; it would just be a bit slower.
> 
> Using a fifo could be an option on unix and turned off on windows.
> 
> > I haven't found an straightforward answer to the lack of SIGUSR1 on windows, this means emerge.go does not compile
> 
> Here, too, it should be possible to make it work without it , just a little bit worse: it would be possible to ask the user to press return when they're done, which would trigger command on the emacs-side that checks whether merging is done and continue.
> 
> I'm sure there's a way of doing IPC on Windows to replace signals, but I have to admit I have zero experience of Windows development, so I can't help with that.
> 
> > Not being a make expert, I added crude alternative for windows: make.bat - which currently skips both of these issues - otherwise Windows/TCP support is mostly there, so perhaps worth merging in for now and I can raise an issue for these two?
> 
> I have nothing against using make.bat under Windows; as long as it works. It should be possible to make make work under both; it seems that Windows sets an env variable "OS" that could be checked.
> 
> Yes, this looks definitely worth merging. Last time, I didn't merge because it affected normal start, but now it only improves the commands, even if support on Windows is not complete yet.

This is really helpful background, thanks alot, will submit a PR to update windows support once my app gets a bit closer to completion, many thanks!
